### ISSUE TITLE
Nullable reference:  no CS8619 when using `new ValueType<T>()`

### ISSUE THREAD

**Issue by dtivel** [dotnet/roslyn-analyzers#4310]
**Version Used**: 
Microsoft Visual Studio Enterprise 2019 Int Preview
Version 16.9.0 Preview 1.0 [30609.169.main]
VisualStudio.16.IntPreview/16.9.0-pre.1.0+30609.169.main
Microsoft .NET Framework
Version 4.8.04084

C# Tools   3.9.0-1.20507.6+4ed928935ce0e5db4327c168dfdad1c820a6a490

**Steps to Reproduce**:

1.  Create a C# .NET Core project.
2.  Add the following code to a new C# file:
```C#
#nullable enable

using System.Threading.Tasks;

interface A
{
    ValueTask<string> B();
}

class C : A
{
    public ValueTask<string> B()
    {
        return D(); // CS8619:  Nullability of reference types in value of type 'ValueTask<string?>' doesn't match target type 'ValueTask<string>'
    }

    private ValueTask<string?> D()
    {
        return new ValueTask<string?>();
    }
}

class D : A
{
    public ValueTask<string> B()
    {
        return new ValueTask<string>(); // No CS8619
    }
}
```

**Expected Behavior**:  Because [`new ValueTask<T>()` returns `default(T)`, which can be `null`, as the result](https://source.dot.net/#System.Private.CoreLib/ValueTask.cs,455), the returned expression in `D.B()` will raise CS8619.

**Actual Behavior**:  No CS8619.

**Comment by stephentoub** [comment]
How is this different from any other similar case with structs?
```C#
#nullable enable

public struct Foo<T>
{
    public T Bar { get; set; }
}

class Z
{
    public void M()
    {
        string s = default(Foo<string>).Bar; // no warning even though Bar is null
    }
}
```

**Comment by sharwell** [comment]
@stephentoub It's not different. It's just outside the scope of the compiler feature and an API-aware analyzer would be able to provide users with actionable inforamtion.

**Comment by stephentoub** [comment]
> It's not different. It's just outside the scope of the compiler feature and an API-aware analyzer would be able to provide users with actionable information.

But specific to `ValueTask<T>`?  Why?

**Comment by sharwell** [comment]
I have no objection to expanding the set of supported types, provided `ValueTask<T>` works. For example, it would be sufficient to treat `ValueTask<T>` as a non-defaultable value type, which forces the use of `new ValueTask<T>(T)`.

**Comment by stephentoub** [comment]
I just don't see what's special about `ValueTask<T>` here.  Structs containing references and arrays are simply two known holes in nullable reference types; I don't know why we'd start adding analyzers on a type-by-type basis to try to patch those.

**Comment by sharwell** [comment]
@dtivel Another option you could employ is an analyzer like [SA1129: Do not use default value type constructor](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1129.md), which is a targeted subset of feature request #1185.

**Comment by dtivel** [comment]
@sharwell, thanks, I'll look at that.

@stephentoub, that "[s]tructs containing references and arrays are simply two known holes in nullable reference types" wasn't known to me.  Does nullable reference types documentation detail these known holes?  I don't recall reading anything about them (like [here](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references)).

**Comment by stephentoub** [comment]
> Does nullable reference types documentation detail these known holes?

I don't know. @BillWagner?

To be clear, what I'm referring to are things like:
```C#
string[] values = new string[42];
string s = values[0]; // s now contains null, no warning
```
and
```C#
struct S
{
    public string Value { get; set; }
}
...
string s = default(S).Value; // s now contains null, no warning
```

The system is not infallible.  You can get both false positives and false negatives.  The goal of the compiler is to help point out places where you could be dereferencing nulls, but by design it both a) misses some cases and b) warns in some cases it needn't.

**Comment by Youssef1313** [comment]
Reading quickly through the [documentation](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references), it doesn't seem to point to this.

This issue is somewhat similar to https://github.com/dotnet/roslyn/issues/47346

**Comment by mavasani** [comment]
I am moving this back to Roslyn repo - this concerns a false positive with a C# compiler nullable warning. There has been a conscious decision to not add diagnostic suppressors for such nullable warnings, and a regular analyzer cannot affect compiler output. Adding an API analyzer that suggests not using first class compiler features does not seem to be a good approach. Regardless, I believe the Roslyn compiler team should decide how to tackle such cases, as there seem to be more then one such scenarios that need to be addressed.
